### ISSUE TITLE
Resource tree slow refresh

### ISSUE THREAD

**Issue by klamkma** [argoproj/argo-cd#8172]
Hello, 

**Describe the bug**

We have a big kubernetes cluster with almost 3000 argocd applications.
Currently we are running ArgoCD 2.2.2.
Since upgrade to version 2 we noticed that refresh of the resource tree for applications is much slower. For example:
I click on "Restart" for a deployment
ReplicaSet appears immediately
New pod appears sometimes after 40 seconds
I've tried increasing values --status-processors, --operation-processors, --kubectl-parallelism-limit for the controller, but it does not help.
Any idea what could we do? Which component is responsible for this refresh, is it argocd-server?

**To Reproduce**

I click on "Restart" for a deployment
ReplicaSet appears immediately
New pod appears sometimes after 40 seconds

**Expected behavior**

Pods should appear faster.

**Version**

```shell
argocd: v2.2.2+03b17e0
  BuildDate: 2022-01-01T06:27:52Z
  GitCommit: 03b17e0233e64787ffb5fcf65c740cc2a20822ba
  GitTreeState: clean
  GoVersion: go1.16.11
  Compiler: gc
  Platform: linux/amd64
```

Thank you.

**Comment by yydzhou** [comment]
We are same here too. Previously we had 4800+ applications the  argocd handles them pretty well, although with some slowness on application listing. After some re-org, we have 3000+ applications now. However, since upgraded v2, the refresh and the sync become very very slow. The refresh action, which supposed to be done in a few seconds can run up to 2 minutes.  The sync waiting is even slower.  compared to previous using experience, I believe there are large room for performance tuning/improvement.

**Comment by alexmt** [comment]
It is really difficult to troubleshoot it remotelly. The controller might be CPU throttled, repo server might need to be scaled up or control plane K8S API server might be slow.

@klamkma , @yydzhou if possible can we have an interactive session (e.g. zoom call) and debug it together. Later we could document changes we've made to help anyone else who faces this issue.

**Comment by yeya24** [comment]
Thank you @alexmt. Would be great to have a debug session together with @yydzhou.

**Comment by klamkma** [comment]
Hello, I'm available for a session too. Thank you @alexmt.

**Comment by klamkma** [comment]
Hi again,

I enabled ARGOCD_ENABLE_GRPC_TIME_HISTOGRAM.
Could you give me some tips how to use it to investigate performance issues?

Thank you

**Comment by leotomas837** [comment]
Any update about this ? We are experiencing the same issue. It may be a duplicate from [this issue](https://github.com/argoproj/argo-cd/issues/8663).

There is enough RAM, CPU, disk space, and we tried multiplying the number of replicas of the controller and the server pods by 4 just to see if it helps, but not at all.

**Comment by jujubetsz** [comment]
I have the same problem. 2.5k Apps, helm,  Argo v2.6.6, 1 very big cluster (HML). I cant see any problem like throttle, OOM's or resource starvation. Did all recommended tunning for high performance. Argocd have a pool with some big nodes just for it to play. Tomorrow i will try to debug the kubernetes cluster to see if the control plane is ok.

**Comment by AnubhavSabarwal** [comment]
Is there any solution for this, we have somewhere around 6000 applications and argocd version is 2.7.2

1. Sync and refresh is very slow
2. Restart and Delete of replicas or deployment doesn't show on ARGO UI.
3. Whenever you delete deployment, pod or replicas always say doesn't exist on ARGOUI

**Comment by klamkma** [comment]
Hi, For us we had a huge improvement in the UI by enabling `--enable-gzip`, but still pod refresh is very slow.

**Comment by jujubetsz** [comment]
@evs-ops, Hi. 

I’ve tried every possible tunning and version in Argocd and got no improvements…. Since my cluster is running in OpenStack/Rancher inside my company cloud, i’m now improving the cluster itself.. Upgrading kubernetes version, etcd performance etc. I’m doing this because i’m seeing lots of timeouts to kubernetes in application-controller and also because none of the tunning worked. This is the logs observed:

`time=“2024-01-08T15:31:23Z” level=info msg=“Failed to watch Deployment.apps on https://x.x.x.x:443: Resyncing Deployment.apps on https://x.x.x.x:443 due to timeout, retrying in 1s” server=“https://kubernetes.default.svc”
time=“2024-01-08T15:34:15Z” level=info msg=“Failed to watch Secret on https://x.x.x.x:443: Resyncing Secret on https://x.x.x.x:443 due to timeout, retrying in 1s” server=“https://kubernetes.default.svc”
time=“2024-01-08T15:35:20Z” level=info msg=“Failed to watch ReplicationController on https://x.x.x.x:443: Resyncing ReplicationController on https://x.x.x.x:443 due to timeout, retrying in 1s” server=“https://kubernetes.default.svc”
`

The symptoms i’m experimenting are:

The navigation in argocd web is fast as expected, but if i delete a pod, for exemple, nothing happens. The box with that pod persist in the argocd frontend, but if i watch the namespace using kubectl the pod is beeing killed and a new pod is beeing scheduled. After several minutes (10 ~15m) the new pod spawn in argocd front. This happen with every object owned by argocd.

**Comment by evs-ops** [comment]
Hi,
Very similar to my problem. I can delete and it would probably take about 5 to 10 min. Refresh no less then 2 min up to 5 min.
It's new to me since in my previous roles I used argo and it was lightning fast :(

**Comment by jujubetsz** [comment]
Hi,

Having more clusters to manage is not a bad thing in my point of view. You can have one replica of application-controller for each cluster. Here are some docs and posts that may help you:

https://www.infracloud.io/blogs/sharding-clusters-across-argo-cd-application-controller-replicas/
https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#argocd-application-controller

Did you tried that?

Another question: Your clusters are managed (GKE, EKS etc) or is the same as me: Self deployed and managed?

**Comment by jujubetsz** [comment]
@evs-ops, 

I bumped my version to [v2.10.0-rc4](https://github.com/argoproj/argo-cd/releases/tag/v2.10.0-rc4) in order to test jitter implementation on reconciliation. You can check the proposal and description [issues/14241](https://github.com/argoproj/argo-cd/issues/14241).

The results till now are incredible, no delay at all in ArgoCD UI/Front. If i delete a pod, the new pod appears instantly, so i recommend you to try if possible. I bumped the version this morning and got an stable environment so far. Will update this thread if something new happens. 

Some general info about my environment:

2.7k Apps
Lots of monorepos, each team/tribe have one ranging from 10 to 200 apps
Only one Cluster
Kubernetes v1.25 running in Openstack/Rancher in private cloud
ArgoCD components have tons of resources to use

![git-requests](https://github.com/argoproj/argo-cd/assets/78041402/2d753c0b-23c3-403e-94e7-4574384ec3cd)
![cpu-usage-total](https://github.com/argoproj/argo-cd/assets/78041402/2bc2cb52-4d31-4303-95cf-4f9b4777393c)
![network-total](https://github.com/argoproj/argo-cd/assets/78041402/3b86c2f8-ee17-43d3-8534-242ed788445b)
![reconciliation-activity](https://github.com/argoproj/argo-cd/assets/78041402/2fcd5b19-1988-4e97-a0f3-a1150a7bb2bb)
![cluster-events](https://github.com/argoproj/argo-cd/assets/78041402/31714670-e12d-463e-a0b8-effa1ef92115)
![reconciliation-performance](https://github.com/argoproj/argo-cd/assets/78041402/9c72fcdf-c038-46c0-a941-121cf2f45233)

**Comment by machine3** [comment]
Have you found the reason?

**Comment by ritheshgm** [comment]
+1 

When running 3,000 applications and engaging in activities such as syncing 200 applications, clicking "Restart" for a deployment immediately displays the ReplicaSet, but new pods may take up to two minutes to appear.

**Comment by machine3** [comment]
Does anyone have any ideas for solving the problem, or a temporary solution?

**Comment by CryptoTr4der** [comment]
Same problem here. The refresh is very slow (~3-5m) per Application. Even with version 2.10.2
1 GIT Repo (monorepo) with ~50 applications
CMP argocd-vault-plugin as sidecar deployed

Tried many things in many hours, but nothing helps atm.

**Comment by gazidizdaroglu** [comment]
+1

When running 3,000 applications and engaging in activities such as syncing 200 applications, clicking "Restart" for a deployment immediately displays the ReplicaSet, but new pods may take up to two minutes to appear.

+1

**Comment by daftping** [comment]
We are encountering a similar issue. In large clusters where Argo CD monitors numerous resources, it is significantly slow in processing watches—taking approximately 7 minutes in our case. Consequently, the Argo CD UI displays outdated information and adversely affects several functionalities that depend on sync waves, such as PruneLast. Eventually, the volume of events from the cluster overwhelmed the system, causing Argo CD to stall completely.

To mitigate this, we disabled tracking of Pods and ReplicaSets, which unfortunately diminishes one of the primary advantages of the Argo CD UI. We also disregarded all irrelevant events and attempted to optimize various settings in the application controller. However, scaling the application controller vertically showed no effect, and horizontal scaling is not feasible for a single cluster due to sharding constraints.

**Comment by CryptoTr4der** [comment]
We have removed all argocd config plugins (switched from argocd-vault-plugin to vault-secrets-webhook) and now everything seems to work smoothly

**Comment by gazidizdaroglu** [comment]
any news? 😢

**Comment by gazidizdaroglu** [comment]
Hey, this thread can help you as well! 

https://cloud-native.slack.com/archives/C01TSERG0KZ/p1721141931660909

**Comment by machine3** [comment]
> Hey, this thread can help you as well!
> 
> https://cloud-native.slack.com/archives/C01TSERG0KZ/p1721141931660909

I'm sorry, I can't access the link you provided. Could you please share some details with me?

**Comment by mpelekh** [comment]
> We are encountering a similar issue. In large clusters where Argo CD monitors numerous resources, it is significantly slow in processing watches—taking approximately 7 minutes in our case. Consequently, the Argo CD UI displays outdated information and adversely affects several functionalities that depend on sync waves, such as PruneLast. Eventually, the volume of events from the cluster overwhelmed the system, causing Argo CD to stall completely.
> 
> To mitigate this, we disabled tracking of Pods and ReplicaSets, which unfortunately diminishes one of the primary advantages of the Argo CD UI. We also disregarded all irrelevant events and attempted to optimize various settings in the application controller. However, scaling the application controller vertically showed no effect, and horizontal scaling is not feasible for a single cluster due to sharding constraints.

We are observing precisely the same issue you described. ArgoCD v2.10.9.
@daftping, did you find a way to resolve the issue without disabling tracking pods and replica sets?

**Comment by andrii-korotkov-verkada** [comment]
The fix is there on master and would be a part of v2.13. It optimizes getting resource tree dfs from `O(<tree_size> * <namespace_resource_count>)` to `O(<namespace_resource_count>)`

**Comment by mpelekh** [comment]
Hi @andrii-korotkov-verkada. Thanks for replying
Do you mean the following fixes?
- https://github.com/argoproj/gitops-engine/commit/6b2984ebc47085852a7b63a0fd0b73c52e986217
- https://github.com/argoproj/argo-cd/commit/267f243a899483fea0a4e6a613c18f62bd342c7e

Thanks for your contribution. IterateHierarchyV2 looks promising.

I actually patched v2.10.9 with the above commits. It helped, but not to the very end.

Even though patches significantly improve performance, Argo CD still can not handle the load from large clusters.

In the screenshot, you can see one of the largest clusters. Here, the patched with the above commits v2.10.9 build is running.
- till 12:50, pods and replica sets are disabled from tracking
- from 12:50 to 13:34, pods and replica sets are enabled to be tracked
- after 13:34, pods and replica sets are disabled from tracking

As can be seen, once pods and rs are enabled to be tracked, the cluster event count falls close to zero, and reconciliation time increases drastically.

![Screenshot 2024-08-09 at 20 40 44](https://github.com/user-attachments/assets/00751664-755c-411d-a108-b432de59844c)

![Screenshot 2024-08-09 at 20 51 00](https://github.com/user-attachments/assets/6ffa97fd-9c66-4282-9e59-b1a88c993567)

Number of pods in cluster: ~76k
Number of rs in cluster: ~52k

@andrii-korotkov-verkada Do you have any ideas on what can be improved?

**Comment by crenshaw-dev** [comment]
Are you hitting CPU throttling?

**Comment by mpelekh** [comment]
No, we don't set CPU limits at all and still have plenty of resources on the node.

We found that the potential reason is lock contention.

Here, I added a few more metrics and found out that when the number of events is significant, sometimes it takes ~5 minutes to acquire a lock, which leads to a delay in reconciliation.
https://github.com/mpelekh/gitops-engine/commit/560ef00bcce9201083200f906f15bf1716fbfcc0#diff-9c9e197d543705f08c9b1bc2dc404a55506cfc2935a988e6007d248257aadb1aR1372

![Screenshot 2024-08-09 at 21 11 33](https://github.com/user-attachments/assets/86255c3d-f8e4-45b1-8898-87f42a3c8b68)

**Comment by andrii-korotkov-verkada** [comment]
I had this attempt https://github.com/argoproj/gitops-engine/issues/602, but the benchmark showed neutral-to-regression in terms of throughput. But maybe average latency can get better, idk.

**Comment by crenshaw-dev** [comment]
I'm curious how much of a performance win you saw from just IterateHierarchy, @mpelekh. Those changes are mostly useful for situations where you have a _ton_ of resources in a single namespace.

Am also super curious if Andrii's locking improvements help with this. If so, that's a strong case for merging those changes.

**Comment by mpelekh** [comment]
> I'm curious how much of a performance win you saw from just IterateHierarchy

@crenshaw-dev The comparison is as follows:

## Large cluster
Number of pods in cluster: ~76k
Number of rs in cluster: ~52k

### v2.10.9 without improvements, only additional metrics are added (deployed at 18:00 according to Grafana charts)
Pods and replica sets are enabled to be watched at 18:25.

![Screenshot 2024-08-09 at 22 42 54](https://github.com/user-attachments/assets/c2ecfef8-575c-4a13-8030-39501c9d29d5)
![Screenshot 2024-08-09 at 22 42 16](https://github.com/user-attachments/assets/54f003dd-197b-4bda-be31-088592fcc4c0)
![Screenshot 2024-08-09 at 22 41 50](https://github.com/user-attachments/assets/8a69cd79-972a-4c0f-b688-706fc652ced1)

### v2.10.9 with improvements (https://github.com/argoproj/gitops-engine/commit/6b2984ebc47085852a7b63a0fd0b73c52e986217 and https://github.com/argoproj/argo-cd/commit/267f243a899483fea0a4e6a613c18f62bd342c7e) and additional metrics (deployed at 21:30 according to Grafana charts)
- till 22:50, pods and replica sets are disabled from tracking
- from 22:50 to 23:34, pods and replica sets are enabled to be tracked
- after 23:34, pods and replica sets are disabled from tracking

![Screenshot 2024-08-09 at 22 57 32](https://github.com/user-attachments/assets/5455dfca-46c4-433e-9ed6-73026c3f26f3)
![Screenshot 2024-08-09 at 22 57 12](https://github.com/user-attachments/assets/2dc9a772-1acf-44c1-9376-b9658e02345e)

Even in the enormous cluster, a very tidy performance improvement can be observed with the v2.10.9 with IterateHierarchyV2 (cluster events count is not completely zero; it's ~3-5k).

 

## Smaller cluster
Here are the results from a smaller cluster (compared to the previous one). Pods and replica sets are watched.
Pods: ~18k
ReplicaSets: ~17k

![Screenshot 2024-08-09 at 22 14 33](https://github.com/user-attachments/assets/ec4f5fa6-ee42-40c0-ab8d-a98b207a37da)
![Screenshot 2024-08-09 at 22 14 44](https://github.com/user-attachments/assets/8ee8a1eb-9157-4676-9ae0-aa530ab55162)
![Screenshot 2024-08-09 at 22 15 00](https://github.com/user-attachments/assets/dba6ee65-6c02-4c7a-afc8-88072cb9a22d)
![Screenshot 2024-08-09 at 22 15 27](https://github.com/user-attachments/assets/d792d371-672a-40ad-9968-9439ed1a54a2)

Before 21:45, the v2.10.9 version from upstream was running.
After 21:45, the patched v2.10.9 version with additional metrics and with the following commits was running.
- https://github.com/argoproj/gitops-engine/commit/6b2984ebc47085852a7b63a0fd0b73c52e986217
- https://github.com/argoproj/argo-cd/commit/267f243a899483fea0a4e6a613c18f62bd342c7e

This is the case when IterateHierarchyV2 improves performance significantly.

**Comment by crenshaw-dev** [comment]
Gotcha. So IterateHierarchy gets us ~90% of the way there, but on a huge cluster we'll still have significant lock contention.

**Comment by mpelekh** [comment]
> Am also super curious if Andrii's locking improvements help with this. If so, that's a strong case for merging those changes.

@crenshaw-dev I am going to create a patched v2.10.9 image with additional metrics and the following fixes:
- https://github.com/argoproj/gitops-engine/commit/6b2984ebc47085852a7b63a0fd0b73c52e986217
- https://github.com/argoproj/argo-cd/commit/267f243a899483fea0a4e6a613c18f62bd342c7e
- https://github.com/argoproj/gitops-engine/pull/603/commits/637dca1a06ff73bb0ac598ac131fedc5a37a60e3

I will share the results once I test it in a large cluster.
FYI @andrii-korotkov-verkada

**Comment by andrii-korotkov-verkada** [comment]
If the pods and replica sets are excluded from tracking, would they not even show up in Argo UI, or would it just make them potentially stale?

**Comment by mpelekh** [comment]
> If the pods and replica sets are excluded from tracking, would they not even show up in Argo UI, or would it just make them potentially stale?

@andrii-korotkov-verkada If the pods and replica sets are excluded from tracking, they are not visible in the ArgoCD UI; only deployment is visible, and nothing underneath.

**Comment by mpelekh** [comment]
> @crenshaw-dev I am going to create a patched v2.10.9 image with additional metrics and the following fixes:
> 
> * [argoproj/gitops-engine@6b2984e](https://github.com/argoproj/gitops-engine/commit/6b2984ebc47085852a7b63a0fd0b73c52e986217)
> * [267f243](https://github.com/argoproj/argo-cd/commit/267f243a899483fea0a4e6a613c18f62bd342c7e)
> * [argoproj/gitops-engine@637dca1](https://github.com/argoproj/gitops-engine/commit/637dca1a06ff73bb0ac598ac131fedc5a37a60e3)
> 
> I will share the results once I test it in a large cluster. FYI @andrii-korotkov-verkada

As we agreed, I tested the patched v2.10.9 build with the following fixes:
- https://github.com/argoproj/gitops-engine/commit/6b2984ebc47085852a7b63a0fd0b73c52e986217
- https://github.com/argoproj/argo-cd/commit/267f243a899483fea0a4e6a613c18f62bd342c7e
- https://github.com/argoproj/gitops-engine/commit/637dca1a06ff73bb0ac598ac131fedc5a37a60e3

### tl;dr
The results are almost the same as with only [IterateHierarchyV2](https://github.com/argoproj/gitops-engine/commit/6b2984ebc47085852a7b63a0fd0b73c52e986217) improvement.
Once the pods and replica sets are enabled for tracking, the Cluster Event Count falls close to zero.
**The logs demonstrate that even though we added the changes that optimized the lock usage, we still have significant lock contention.**

### Details
The patched image has been deployed to one of the most largest clusters, where the pods and replica sets are disabled from tracking.

![Screenshot 2024-08-13 at 15 38 23](https://github.com/user-attachments/assets/93a57d36-bd39-4cb8-bf52-cee8e37ec795)

Please take a look at where one of the additional logs has been added - https://github.com/mpelekh/gitops-engine/blob/e773bed14ca188333ce5f3aa9ca08ab582eff360/pkg/cache/cluster.go#L1429.
This log shows how much time it takes to acquire a lock.

The results are as follows.

#### Pods and ReplicaSets are disabled from tracking
```
time to gather logs - from 2024-08-12T15:13:10Z to 2024-08-12T15:24:21Z
total number of processed events during that time - 76482
from 0ms to 1000ms - 75678
from 1000ms to 10000ms - 795
from 10000ms to 20000ms - 9
from 20000ms to 30000ms - 0
from 30000ms to 40000ms - 0
from 40000ms to 50000ms - 0
from 50000ms to 60000ms - 0
from 60000ms and higher - 0
```

#### Enable ReplicaSets to be tracked. The pods are still excluded.
APIs count increased to 85
Resource count became ~130k

```
time to gather logs - from 2024-08-12T15:25:41Z to 2024-08-12T15:52:27Z
total number of processed events during that time - 123501
from 0ms to 1000ms - 120403
from 1000ms to 10000ms - 3085
from 10000ms to 20000ms - 4
from 20000ms to 30000ms - 3
from 30000ms to 40000ms - 6
from 40000ms to 50000ms - 0
from 50000ms to 60000ms - 0
from 60000ms and higher - 0
```

#### Include ReplicaSets and Pods for watching
![Screenshot 2024-08-12 at 19 24 29](https://github.com/user-attachments/assets/99cefe59-f2db-4b3d-99bd-cd0c17009794)

```
time to gather logs - from 2024-08-12T16:18:58Z to 2024-08-12T16:38:10Z
total number of processed events during that time - 14856
from 0ms to 1000ms -    11006
from 1000ms to 10000ms - 3812
from 10000ms to 20000ms - 8
from 20000ms to 30000ms - 15
from 30000ms to 40000ms - 1
from 40000ms to 50000ms - 5
from 50000ms to 60000ms -  4
from 60000ms and higher - 5
```

We see that the number of processed events decreased significantly when the ReplicaSets and Pods were included for watching.
The logs demonstrate that even though we added the [changes that optimized the lock usage](https://github.ihs.demonware.net/mpelekh/gitops-engine/commit/e773bed14ca188333ce5f3aa9ca08ab582eff360#diff-9c9e197d543705f08c9b1bc2dc404a55506cfc2935a988e6007d248257aadb1aR1429), we still have significant lock contention.

Do you have any thoughts regarding how we can better optimize the lock usage in the cluster so as to handle such a huge number of resources (~210k when RS and Pods are included)?

FYI @crenshaw-dev @andrii-korotkov-verkada

**Comment by mpelekh** [comment]
Has anyone tried changing the global lock approach by fine-grained locking to avoid lock contentions?

**Comment by andrii-korotkov-verkada** [comment]
I don't have thoughts how to optimize lock usage unfortunately. The approach I'm learning about now is a cell architecture with multiple clusters https://github.com/argoproj/argo-cd/discussions/19607.

**Comment by Ga13Ou** [comment]
We are experiencing slowness issue on our bigger clusters as well. It would be really helpful for debugging if those temporary lock metrics are added to the metrics exported by Argo

**Comment by vd-arnaud** [comment]
@mpelekh sorry for dumb question but how do you disable pod/replicaset tracking? I'm having same issue as descrbed and I would to try this

**Comment by andrii-korotkov-verkada** [comment]
Version 2.13.0 has optimizations mentioned. @klamkma, can you upgrade and let us know if it solves the issue, please?

**Comment by klamkma** [comment]
@andrii-korotkov-verkada, your PRs are still open:
https://github.com/argoproj/argo-cd/pull/20329
https://github.com/argoproj/gitops-engine/pull/629

Are you sure there is any fix in version  2.13.0?

Thanks!

**Comment by andrii-korotkov-verkada** [comment]
Lock contention fix idea didn't work, but another idea with a more optimal tree construction did.

**Comment by andrii-korotkov-verkada** [comment]
The linked PRs are not mine though.

**Comment by mpelekh** [comment]
@klamkma, The following fixes from @andrii-korotkov-verkada resolve the issue for most cases, and they are part of the v2.13 release:
- https://github.com/argoproj/gitops-engine/commit/6b2984ebc47085852a7b63a0fd0b73c52e986217
- https://github.com/argoproj/argo-cd/commit/267f243a899483fea0a4e6a613c18f62bd342c7e

But on huge clusters, we still have significant lock contention, which leads to performance issues.
The following PRs aim to resolve the performance issue on large clusters not observed on smaller clusters. More details about it here - https://github.com/argoproj/gitops-engine/pull/629#issuecomment-2393085365 
- https://github.com/argoproj/gitops-engine/pull/629
- https://github.com/argoproj/argo-cd/pull/20329

These PRs are still open, and I am focused on making them ready to be merged.

**Comment by vladst3f** [comment]
hi @mpelekh, any ETA on when they'd get merged ? cheers

**Comment by mpelekh** [comment]
> hi @mpelekh, any ETA on when they'd get merged ? cheers

@vladst3f Thanks for reaching out! I'm actively working on it and am interested in getting it merged. I'll keep you updated and aim to have it ready soon. I appreciate your patience and enthusiasm!
